mountpoint-s3-crt/src/s3/buffer.rs

```diff
@@ -0,0 +1,136 @@
+//! Data buffers used by the S3 Client.
+
+use std::ptr::NonNull;
+use std::{mem::ManuallyDrop, ops::Deref};
+
+use mountpoint_s3_crt_sys::{
+    aws_byte_cursor, aws_s3_buffer_ticket, aws_s3_buffer_ticket_acquire, aws_s3_buffer_ticket_release,
+};
+
+use crate::aws_byte_cursor_as_slice;
+
+/// A slice of a data buffer.
+///
+/// Wrapper for a [aws_byte_cursor] pointing to (part of) the body returned by
+/// a meta request. It can optionally hold a reference to a [BufferTicket] which
+/// can be used to acquire ownership of the underlying memory pool buffer.
+#[derive(Debug)]
+pub struct Buffer<'slice> {
+    slice: &'slice aws_byte_cursor,
+    ticket: Option<&'slice BufferTicket>,
+}
+
+impl<'slice> Buffer<'slice> {
+    /// Create a new instance from a [aws_byte_cursor] and optional associated [aws_s3_buffer_ticket].
+    ///
+    /// SAFETY: slice must be a valid `aws_byte_cursor`.
+    pub(crate) fn new_unchecked(slice: &'slice aws_byte_cursor, ticket: Option<&'slice BufferTicket>) -> Self {
+        Self { slice, ticket }
+    }
+
+    /// Acquire ownership of the underlying memory pool buffer and return an [OwnedBuffer].
+    ///
+    /// Fails and returns [None] if no [BufferTicket] was provided.
+    pub fn to_owned_buffer(&self) -> Option<OwnedBuffer> {
+        let ticket = self.ticket?.clone();
```

Comment:
may be a comment here, that we acquire ownership / call `aws_s3_buffer_ticket_acquire` here